File: mobile/shared/src/commonMain/kotlin/uk/foodclassifier/text/OcrTextQuality.kt
package uk.foodclassifier.text

/**
 * Heuristic score for comparing OCR passes — higher means more likely a full ingredient list.
 */
internal fun scoreOcrText(text: String): Int {
    if (text.isBlank()) {
        return 0
    }
    var score = text.length
    score += text.count { it == ',' } * 25
    score += text.count { it == '(' || it == '[' } * 15
    if (Regex("""(?i)\bingredients\b""").containsMatchIn(text)) {
        score += 120
    }
    if (Regex("""(?i)\b(?:skyr|yogurt|milk|concentrate|starch|flavouring|sweetener)\b""").containsMatchIn(text)) {
        score += 40
    }
    return score
}

internal fun pickBetterOcrText(vararg candidates: String): String {
    return candidates
        .filter { it.isNotBlank() }
        .maxByOrNull { scoreOcrText(it) }
        .orEmpty()
}

File: flutter/lib/text/ocr_text_quality.dart
import 'regex_utils.dart';

int scoreOcrText(String text) {
  if (text.trim().isEmpty) {
    return 0;
  }
  var score = text.length;
  score += ','.allMatches(text).length * 25;
  score += RegExp(r'[(\[]').allMatches(text).length * 15;
  if (ic(r'\bingredients\b').hasMatch(text)) {
    score += 120;
  }
  if (ic(r'\b(?:skyr|yogurt|milk|concentrate|starch|flavouring|sweetener)\b')
      .hasMatch(text)) {
    score += 40;
  }
  return score;
}

String pickBetterOcrText(List<String> candidates) {
  return candidates
      .where((candidate) => candidate.trim().isNotEmpty)
      .fold<String?>(
        null,
        (best, candidate) {
          if (best == null || scoreOcrText(candidate) > scoreOcrText(best)) {
            return candidate;
          }
          return best;
        },
      ) ??
      '';
}
